libzecale/circuits/pairing/bls12_377_pairing.hpp

```diff
@@ -325,6 +325,119 @@ class bls12_377_ate_miller_loop_gadget : public libsnark::gadget<libff::Fr<ppT>>
     void generate_r1cs_witness();
 };
 
+template<typename ppT>
+class bls12_377_final_exp_first_part_gadget
+    : public libsnark::gadget<libff::Fr<ppT>>
+{
+public:
+    using FieldT = libff::Fr<ppT>;
+    using FqkT = libff::Fqk<libsnark::other_curve<ppT>>;
+
+    // Follows the implementation used in
+    // libff::bls12_377_final_exponentiation_first_chunk() (see
+    // libff/algebra/curves/bls12_377/bls12_377_pairing.cpp), which in turn
+    // follows:
+    //   https://eprint.iacr.org/2016/130.pdf
+
+    Fp12_2over3over2_variable<FqkT> _in;
+    Fp12_2over3over2_variable<FqkT> _result;
+
+    // A = elt^(q^6)
+    // B = elt^(-1)
+    Fp12_2over3over2_inv_gadget<FqkT> _B;
+    // C = A * B = elt^(q^6 - 1)
+    Fp12_2over3over2_mul_gadget<FqkT> _C;
+    // D = C^(q^2) = elt^((q^6 - 1) * (q^2))
+    // result = D * C = elt^((q^6 - 1) * (q^2 + 1))
+    Fp12_2over3over2_mul_gadget<FqkT> _D_times_C;
+
+    bls12_377_final_exp_first_part_gadget(
+        libsnark::protoboard<FieldT> &pb,
+        const Fp12_2over3over2_variable<FqkT> &in,
+        const Fp12_2over3over2_variable<FqkT> &result,
+        const std::string &annotation_prefix);
+
+    const Fp12_2over3over2_variable<FqkT> &result() const;
+    void generate_r1cs_constraints();
+    void generate_r1cs_witness();
+};
+
+template<typename ppT>
+class bls12_377_exp_by_z_gadget : public libsnark::gadget<libff::Fr<ppT>>
+{
+public:
+    using FieldT = libff::Fr<ppT>;
+    using FqkT = libff::Fqk<libsnark::other_curve<ppT>>;
+    using cyclotomic_square = Fp12_2over3over2_cyclotomic_square_gadget<FqkT>;
+    using multiply = Fp12_2over3over2_mul_gadget<FqkT>;
+    using unitary_inverse = Fp12_2over3over2_cyclotomic_square_gadget<FqkT>;
+
+    Fp12_2over3over2_variable<FqkT> _in;
+    Fp12_2over3over2_variable<FqkT> _result;
+    std::vector<std::shared_ptr<cyclotomic_square>> _squares;
+    std::vector<std::shared_ptr<multiply>> _multiplies;
+    std::shared_ptr<unitary_inverse> _inverse;
+
+    bls12_377_exp_by_z_gadget(
+        libsnark::protoboard<FieldT> &pb,
+        const Fp12_2over3over2_variable<FqkT> &in,
+        const Fp12_2over3over2_variable<FqkT> &result,
+        const std::string &annotation_prefix);
+
+    const Fp12_2over3over2_variable<FqkT> &result() const;
+    void generate_r1cs_constraints();
+    void generate_r1cs_witness();
+
+private:
+    void initialize_z_neg(libsnark::protoboard<FieldT> &, const std::string &);
+    void initialize_z_pos(libsnark::protoboard<FieldT> &, const std::string &);
```

Comment:
What about adding arg names here?, i.e. `pb` and `str`

Reply:
added args